tests/layer/test_sort_pooling.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import pytest
+import numpy as np
+from stellargraph.layer.sort_pooling import SortPooling
+
+
+def test_sorting_padding():
+
+    data = np.array([[3, 4, 0], [1, 2, 2], [5, 0, 1]], dtype=int).reshape((1, 3, 3))
+    mask = np.array([[True, True, True]])
+    data_sorted = np.array(
+        [[1, 2, 2], [5, 0, 1], [3, 4, 0], [0, 0, 0]], dtype=int
+    ).reshape((1, 4, 3))
+
+    layer = SortPooling(k=4)
+
+    data_out = layer([data, mask])
+
+    assert np.array_equal(data_out, data_sorted)
+
+    # for mini-batch of size > 1
+    data = np.array([[3, 1], [1, 2], [5, 0], [0, -4]], dtype=int).reshape((2, 2, 2))
+    mask = np.array([[True, True], [True, True]])
+    data_sorted = np.array(
+        [[1, 2], [3, 1], [0, 0], [5, 0], [0, -4], [0, 0]], dtype=int
+    ).reshape((2, 3, 2))
+
+    layer = SortPooling(k=3)
+
+    data_out = layer([data, mask])
+
+    assert np.array_equal(data_out, data_sorted)
+
+
+def test_sorting_truncation():
+    data = np.array([[3, 4, 0], [1, 2, 2], [5, 0, 1]], dtype=int).reshape((1, 3, 3))
+    mask = np.array([[True, True, True]])
+
+    data_sorted = np.array([[1, 2, 2], [5, 0, 1]], dtype=int).reshape((1, 2, 3))
+
+    layer = SortPooling(k=2)
+
+    data_out = layer([data, mask])
+
+    assert np.array_equal(data_out, data_sorted)
+
+    # for mini-batch of size > 1
+    data = np.array([[3, 1], [1, 2], [5, 0], [0, -4]], dtype=int).reshape((2, 2, 2))
+    mask = np.array([[True, True], [True, True]])
+
+    data_sorted = np.array([[1, 2], [5, 0]], dtype=int).reshape((2, 1, 2))
+
+    layer = SortPooling(k=1)
+
+    data_out = layer([data, mask])
+
+    assert np.array_equal(data_out, data_sorted)
+
+
+def test_sorting_negative_values():
+
+    data = np.array([[3, 4, 0], [1, 2, -1], [5, 0, 1]], dtype=int).reshape((1, 3, 3))
+    mask = np.array([[True, True, True]])
+
+    data_sorted = np.array([[5, 0, 1], [3, 4, 0], [1, 2, -1]], dtype=int).reshape(
+        (1, 3, 3)
+    )
+
+    layer = SortPooling(k=3)
+
+    data_out = layer([data, mask])
+
+    assert np.array_equal(data_out, data_sorted)
+
+
+def test_mask():
+    data = np.array([[3, 4, 0], [1, 2, -1], [5, 0, 1]], dtype=int).reshape((1, 3, 3))
+    mask = np.array([[True, True, True]])
```

Comment:
If this test is testing the mask, should this be something like:

```suggestion
    mask = np.array([[True, True, False]])
```

otherwise this seems like this first case is the same as `test_sorting_truncation`?

Reply:
This is testing the case that the mask allows all rows to be sorted and then truncated correctly because `k=2`. In line 91, the last row [5, 0, 1] is correctly moved to the first row followed by [3, 4, 0] while the last element is truncated from the result. If the mask is changed as suggested, then the result will be [3, 4, 0] followed by [1, 2, -1]. Generally, I want to make sure here that if all mask values are True, sorting still works across all rows.

There is a test for `mask=[True, True, False]` after this one.

Reply:
Makes sense, I guess I am just a bit disconcerted by the complete redundancy with the `test_sorting_truncation` test above, since that test does exactly the same process just with slightly different input values. My question was whether this was a typo for intending to test the combination of a non-trivial mask and truncation (and the next one is the combination of a non-trivial mask + padding), since all the other tests in this file test various combinations with a trivial mask.